src/vsr/replica.zig

```diff
@@ -8313,6 +8566,80 @@ pub fn ReplicaType(
+                assert(self.journal.status == .init);
+            }
+
+            if (self.release < release_target) {
+                // Upgrading to new release.
+                // We checkpointed or state-synced an upgrade.
+                //
+                // Even though we are upgrading, our target version is not necessarily available in
+                // our binary. (In this case, release_execute() is responsible for error-ing out.)
+                maybe(self.release == self.releases_bundled.get(0));
+                assert(self.commit_min == self.op_checkpoint() or
+                    self.commit_min == vsr.Checkpoint.trigger_for_checkpoint(self.op_checkpoint()));
+                maybe(self.journal.status == .init);
+            }
+
+            log.info("{}: release_transition: release={}..{} (reason={s})", .{
+                self.replica,
+                self.release,
+                release_target,
+                source.fn_name,
+            });
+
+            self.release_execute(self, release_target);
+            // At this point, depending on the implementation of release_execute():
+            // - For testing/cluster.zig: `self` is no longer valid – the replica has been
+            //   deinitialized and re-opened on the new version.
+            // - For tigerbeetle/main.zig: This is unreachable (release_execute() will not return).
+        }
+
+        /// Returns the next checkpoint's `CheckpointState.release`.
+        fn release_for_next_checkpoint(self: *const Self) u16 {
+            assert(self.commit_stage != .idle);
+            assert(self.commit_min == self.op_checkpoint_next_trigger());
+            assert(self.release ==
+                self.superblock.working.vsr_state.checkpoint.release);
+
+            var found_upgrade: usize = 0;
+            for (self.op_checkpoint_next() + 1..self.op_checkpoint_next_trigger() + 1) |op| {
+                const header = self.journal.header_for_op(op).?;
+                assert(header.operation != .reserved);
+
+                if (header.operation == .upgrade) {
+                    found_upgrade += 1;
+                } else {
+                    // Only allow the next checkpoint's release to advance if the entire last bar
+                    // preceding the checkpoint trigger consists of operation=upgrade.
+                    //
+                    // Otherwise we would risk the following:
+                    // 1. Execute op=X in the state machine on version v1.
+                    // 2. Upgrade, checkpoint, restart.
+                    // 3. Replay op=X when recovering from checkpoint on v2.
+                    // If v1 and v2 produce different results when executing op=X, then an assertion
+                    // will trip (as v2's reply doesn't match v1's in the client sessions).
+                    assert(found_upgrade < constants.lsm_batch_multiple);
+                    maybe(found_upgrade > 0);
```

Comment:
To explain this maybe: if a cluster is idle, the op is between a checkpoint and its trigger, and the cluster receives an upgrade request, then:

* we want to append the request to log immediately
* but we want to pulse a checkpoint worth of upgrades up to the _next_ checkpoint, because the first critical bar of the current checkpoint is already poisoned by a non-upgrade request.

Reply:
Wait, no something doesn't add up here? If there is an upgrade committed, anything on top of it should also be an upgrade? 

That is, across the _entire_ loop, `found_upgrade` could be greater than 0 but less than `lsm_batch_multiple`. However, if there's a non-upgrade op in this batch, it has to be the first op, no? 

Am I misunderstanding the big picture here?

Reply:
Ah, you're right, this should be
```
assert(found_upgrade == 0);
```

This is a relic from an earlier iteration which allowed non-upgrades to queue up behind the upgrades if no upgrade had actually committed yet.